Beta2: Semantic UAST

juanjux opened the pull request:
This release adds all semantic objects to the C++ driver except `Import` which will need the handing/merging of the preprocesor nodes (like `#import`) that CDT provides on a separate list (which will be the beta3/RC).

Remaining problem: function's (`FunctionGroup`) variable arguments are not correctly handed. The problem being that CDT doesn't give a "..." argument with a "IsVarArg" or similar property like the other drivers but instead puts that property in the parent function node. But then the `Each()`, used to handle each argument, doesn't seem to generate a valid `Array` so I can't append the synthetic "..." argument to it using `Append()`. I'm investigating how to do it but any help (@dennwc) would be appreciated :)

dennwc commented on fixtures/anonymous_class_declaration.cpp.sem.uast:
```diff
@@ -82,13 +80,10 @@
                      },
-                     IsAlternate: false,
-                     IsOverloadedOperator: false,
+                     Name: "",
```
Hmm, that's weird. I think we need to wipe this kind of nodes from the Semantic tree. Should be an empty array in this case.

juanjux replied:
About the `falses` being removed, I changed the driver in this version to only show the "Is..." properties if true because they're a lot and make the AST too big. About the ImplicitNames with the Name to empty, they're mostly duplicated of other properties and actually there is a switch for CDT to not visit those "ImplicitNames", I'll check the fixtures to see if they give some information that other nodes doesn't and if that's not the case remove them.

creachadair replied:
One thing to look out for is that if you don't specify the attributes, the reader may not be able to tell the difference between "this node has `Property:true`" and "`Property` is not relevant to this node". 

In Go, we get away with zero-values because the struct provides a schema. But in a bag-of-props model, it's not intrinsically obvious which attributes _should_ be there (but are defaulted to empty/false/zero). Whether we can get away with it in UAST probably depends on whether the reader knows which attributes go with each node.

juanjux replied:
A lot are pretty language specific (IsVolatile, IsFriend...), but, apart from this drawback, I discovered another one in the previous beta: it complicates annotations mightily by forcing the use of `FieldRoles{Optional...}}` on almost every node. 

So I'm not sure if the reduced AST size is worthy considering all this. I'll disable it for beta3 (that I'll PR today) and give me your opinion there over the generated fixtures.

dennwc replied:
This comment was reffering to `Name: ""` in the `Identifier` -  that should never happen.
False values are a different (but valid) issue.

dennwc commented on fixtures/arithmeticops.cpp.sem.uast:
```diff
@@ -28,1429 +27,44 @@
-                        IsLValue: false,
-                        kind: "integer_constant",
-                     },
+                     Statements: [],
```
It looks like the whole function body was discarded. Please diff `.uast` vs `.sem.uast` before approving it as a correct fixture test file.

juanjux replied:
Yes, I'm investigating it, I was so focused on getting the parameters right that I didn't check the body :/.

juanjux replied:
Should be fixed now.

dennwc commented on fixtures/array_initilizer.cpp.sem.uast:
```diff
@@ -107,14 +105,10 @@
-                     IsLValue: true,
-                     kind: "string_literal",
+                     Format: "",
+                     Value: "\"abc\"",
```
Should be unescaped in Semantic (`"abc"`)

juanjux replied:
Fixed, thanks.

juanjux commented:
Added @tsolakoua new benchmark fixture (from ~master) and generated the `.native`, `.uast` and `.sem.uast` files. Also fixed an issue with empty `CPPASTName` nodes. Should be ready for review again.

juanjux commented:
@dennwc PTAL when you can, I'm about to publish `beta3` so either me merge this (with the `FunctionGroup` varargs problem since it's not critical) or close as superseded by the new one, but then it would be doubly huge compared to ~master so I would prefer to merge this first.

juanjux commented:
@dennwc for beta3 I've enabled the option to not visit implicit names since all I checked were either empty or the same as the `Name` and all the usages of this class I've seen (including IBM's own) disabled them so this shouldn't be an issue anymore.
